core/trino-main/src/test/java/io/trino/block/BlockAssertions.java

```diff
@@ -92,6 +114,148 @@ public static void assertBlockEquals(Type type, Block actual, Block expected)
+            }
+            else {
+                // RowType doesn't need offsets, so we just use 1,
+                // for ArrayType and MapType we choose randomly either array length or map size at the current position
+                offsets[position + 1] = offsets[position] + (type instanceof RowType ? 1 : RANDOM.nextInt(ENTRY_SIZE) + 1);
+            }
+        }
+
+        // Builds the nested block of size offsets[positionCount].
+        if (type instanceof ArrayType) {
+            Block valuesBlock = createRandomBlockForType(((ArrayType) type).getElementType(), offsets[positionCount], nullRate);
+            return fromElementBlock(positionCount, Optional.ofNullable(isNull), offsets, valuesBlock);
+        }
+        if (type instanceof MapType) {
+            MapType mapType = (MapType) type;
+            Block keyBlock = createRandomBlockForType(mapType.getKeyType(), offsets[positionCount], 0.0f);
+            Block valueBlock = createRandomBlockForType(mapType.getValueType(), offsets[positionCount], nullRate);
+
+            return mapType.createBlockFromKeyValue(Optional.ofNullable(isNull), offsets, keyBlock, valueBlock);
+        }
+        if (type instanceof RowType) {
+            List<Type> fieldTypes = type.getTypeParameters();
+            Block[] fieldBlocks = new Block[fieldTypes.size()];
+
+            for (int i = 0; i < fieldBlocks.length; i++) {
+                fieldBlocks[i] = createRandomBlockForType(fieldTypes.get(i), positionCount, nullRate);
+            }
+
+            return RowBlock.fromFieldBlocks(positionCount, Optional.ofNullable(isNull), fieldBlocks);
+        }
+
+        throw new IllegalArgumentException(format("type %s is not supported.", type));
+    }
+
+    public static Block createRandomBooleansBlock(int positionCount, float nullRate)
+    {
+        return createBooleansBlock(generateListWithNulls(positionCount, nullRate, RANDOM::nextBoolean));
+    }
+
+    public static Block createRandomIntsBlock(int positionCount, float nullRate)
+    {
+        return createIntsBlock(generateListWithNulls(positionCount, nullRate, RANDOM::nextInt));
+    }
+
+    public static Block createRandomLongDecimalsBlock(int positionCount, float nullRate)
+    {
+        return createLongDecimalsBlock(generateListWithNulls(
+                positionCount,
+                nullRate,
+                () -> BigDecimal.valueOf(RANDOM.nextDouble()).multiply(BigDecimal.valueOf(RANDOM.nextInt())).toString()));
+    }
+
+    public static Block createRandomLongsBlock(int positionCount, float nullRate)
+    {
+        return createLongsBlock(generateListWithNulls(positionCount, nullRate, RANDOM::nextLong));
+    }
+
+    public static Block createRandomSmallintsBlock(int positionCount, float nullRate)
+    {
+        return createTypedLongsBlock(
+                SMALLINT,
+                generateListWithNulls(positionCount, nullRate, () -> RANDOM.nextLong() % Short.MIN_VALUE));
```

Comment:
`RANDOM.nextLong() % Short.MIN_VALUE` just cast it to `(short)`?

Reply:
the `createTypedLongsBlock` is expecting `long` so would have to cast to `long` again

Reply:
Cast from `long` to short is implicit